### ISSUE TITLE
Scaling

### ISSUE THREAD

**Issue by knutole** [mapic/mapic#22]
# Scaling

## How
There are several things we can do to improve processing speed: 
1. Scale across nodes
2. Automate tile creation (utilize downtime and create tiles before they are requested by end-user).
3. Vector tiles

Current setup explained [here](https://github.com/mapic/mapic/issues/23).

### Scale across nodes
The unit doing the heavy lifting is [Mile](https://github.com/mapic/mile/tree/master/src), the tile server (together with PostGIS serving data). These are the containers which need to be scaled/replicated across hardware nodes. This will be implemented as a MVP, see below.

### Automate tile creation
Raster tiles are currently created upon request only (by browsing in map). By implementing a function (and button in GUI) for creating all tiles before they are requested, the experience for the end-user is that tiles are served immediately. This is part of post-MVP.

### Vector tiles
By creating vector tiles, the tile-set needs to be created only once, even if stylesheet changes. This cuts the needed processing work by orders of magnitude. This is not part of this MVP.

### Goal: 
Scale raster tile processing across 3 nodes.

### Entails:
- Utilize more nodes (=servers/hardware) to host `mile` and `postgis` containers.
- Make containers standalone and agnostic.
- Load-balancer which redirects tile requests to one of several `mile` containers. (`nginx -> load-balancer -> mile`)
- Internal addressing for available containers (mile, postgis, redis).
- Handling of `access_tokens` across nodes.
- Refactoring of data flow.
- Replicating `postgis` containers (ie. with sync). (This will likely create unforeseen sync problems.) [Example.](https://gist.github.com/volkanunsal/ad2173e2649393fcd3b6)
- Deploy with replicated nodes. (Adding/removing containers to running server is post-MVP.)
- 1-3 `postgis`, 1-3 `mile` nodes. 1 node to host the rest (Redis, Mongo, Engine).
- Orchestrated with Docker Swarm mode and overlay networks.
    - `Master (node 1)`: Engine, webserver, mongo, redis, PostGIS (for importing data)
    - `Slaves (nodes 2-4)`: Mile, redis-layers-replicas
    - `Slaves (nodes 2-4)`: PostGIS-replicas (for exporting data)
- Result: 3 x 12 cpu = 36 parallell threads processing raster tiles. 

### Schematic
What the server/docker infrastructure will look like in MVP:

![server-infrasctructure](https://cloud.githubusercontent.com/assets/2197944/26045359/1e1b2364-394a-11e7-90c4-6a8bc2c00848.png)

> _Notes to schematic:_
> - _Some connections are not shown, for example `mile` connections to redis/mongo/engine._
> - _PostGIS containers in `nodes 2-4` replicate with "master" PostGIS on `node 1`._
> - _PostGIS on `node 1` is for `importing` only. PostGIS on `nodes 2-4` are for `exporting` data only._
> - _Redis containers will probably be present on `nodes 2-4` as well, replicating with `node 1`._
> - _"Load-balancer" will probably be a standalone NodeJS proxy with knowledge of available resources._
> 

### Priorities
More or less in order of importance:
- Load-balancing based on availability instead of random/even. Weight calc based on current workload (number * size), and available cpu's on containers - for both `postgis` and `mile`.
- Automatic service discovery, allowing spinning containers up and down at will.
- Automate tile creation. Processing all zoom levels on request. Scheduling of processing when there's availability. 
- Refactoring of databases (redis, mongo). Make all processes talk to one API. 
- Refactoring of `mile` connections to redis/mongo/api.
- Explore parallell reading from PostGIS (?). Currently reading with one thread. 
- Stability improvements, optimizations.

**Comment by knutole** [comment]
Moved to https://github.com/mapic/mapic/issues/36
